In [1]:
import SimpleITK as sitk
import matplotlib.pyplot as plt
import numpy as np

import pymia.filtering.postprocessing as pymia_postproc
import pymia.filtering.filter as pymia_fltr
from matplotlib.widgets import Slider

%matplotlib qt


In [2]:
# Replace 'your_image.mha' with the actual file path
image_path = 'mia-result/2024-10-30-08-50-27/117122_SEG.mha'
image = sitk.ReadImage(image_path)

image_path = 'mia-result/2024-10-30-08-50-27/117122_GT.mha'
GT = sitk.ReadImage(image_path)

# Convert the image to a NumPy array if needed
image_array = sitk.GetArrayFromImage(image)
GT_array = sitk.GetArrayFromImage(GT)

# Print the image dimensions
print(f"Image size: {image.GetSize()}")
print(f"Image array shape: {image_array.shape}")
print('-'*20)
print(f"GT size: {GT.GetSize()}")
print(f"GT array shape: {GT_array.shape}")

Image size: (197, 233, 189)
Image array shape: (189, 233, 197)
--------------------
GT size: (197, 233, 189)
GT array shape: (189, 233, 197)


# Image

In [3]:
# Set up the plot for the original image slices
max_slice = image_array.shape[0] - 1
fig, ax = plt.subplots(figsize=(8, 8))
fig.suptitle("Original Image Slices")

# Initialize the initial slice index to the middle slice
initial_slice_index = max_slice // 2

# Create a color map using 'tab20'
cmap = plt.get_cmap('tab20')

# Function to display the original image slice with a tab20 colormap
def display_slice(slice_index):
    ax.clear()
    # Normalize the image to range [0, 1] for colormap application
    normalized_image = (image_array[slice_index] - np.min(image_array[slice_index])) / (np.max(image_array[slice_index]) - np.min(image_array[slice_index]))
    # Map the normalized image to colormap
    colored_image = cmap(normalized_image)  # Get colors from the colormap
    ax.imshow(colored_image)  # Display the colored image
    ax.set_title(f"Original Image Slice {slice_index}")
    ax.axis("off")
    fig.canvas.draw_idle()

# Display the initial slice
display_slice(initial_slice_index)

# Add a slider below the plot
ax_slider = plt.axes([0.2, 0.05, 0.6, 0.03], facecolor='lightgoldenrodyellow')
slice_slider = Slider(ax_slider, 'Slice', 0, max_slice, valinit=initial_slice_index, valstep=1)

# Update the displayed slice when the slider is changed
def update(val):
    slice_index = int(slice_slider.val)
    display_slice(slice_index)

slice_slider.on_changed(update)

plt.show()

# GT

In [4]:
# Set up the plot for the original image slices
maGTce = image_array.shape[0] - 1
fig, ax = plt.subplots(figsize=(8, 8))
fig.suptitle("GT Image Slices")

# Initialize the initial slice index to the middle slice
initial_slice_index = max_slice // 2

# Create a color map using 'tab20'
cmap = plt.get_cmap('tab20')

# Function to display the original image slice with a tab20 colormap
def display_slice(slice_index):
    ax.clear()
    # Normalize the image to range [0, 1] for colormap application
    normalized_image = (GT_array[slice_index] - np.min(GT_array[slice_index])) / (np.max(GT_array[slice_index]) - np.min(GT_array[slice_index]))
    # Map the normalized image to colormap
    colored_image = cmap(normalized_image)  # Get colors from the colormap
    ax.imshow(colored_image)  # Display the colored image
    ax.set_title(f"GT Image Slice {slice_index}")
    ax.axis("off")
    fig.canvas.draw_idle()

# Display the initial slice
display_slice(initial_slice_index)

# Add a slider below the plot
ax_slider = plt.axes([0.2, 0.05, 0.6, 0.03], facecolor='lightgoldenrodyellow')
slice_slider = Slider(ax_slider, 'Slice', 0, max_slice, valinit=initial_slice_index, valstep=1)

# Update the displayed slice when the slider is changed
def update(val):
    slice_index = int(slice_slider.val)
    display_slice(slice_index)

slice_slider.on_changed(update)

plt.show()

# Perform postprocessing

### Erode - FillHole - Dilate

In [5]:
# Assuming image has been defined and processed (eroded, filled, dilated)

# Define the image processing pipeline
erode_filter = sitk.BinaryErodeImageFilter() 
erode_filter.SetForegroundValue(1) 
erode_filter.SetKernelType(sitk.sitkBall) 
eroded_image = erode_filter.Execute(image)
eroded_image_array = sitk.GetArrayFromImage(eroded_image)

fillhole_filter = sitk.BinaryFillholeImageFilter() 
fillhole_filter.SetFullyConnected(True) 
fillhole_filter.SetForegroundValue(1) 
filled_image = fillhole_filter.Execute(eroded_image) 
filled_image_array = sitk.GetArrayFromImage(filled_image)

dilate_filter = sitk.BinaryDilateImageFilter() 
dilate_filter.SetForegroundValue(1) 
dilate_filter.SetKernelType(sitk.sitkBall)
dilated_image = dilate_filter.Execute(filled_image) 
dilated_image_array = sitk.GetArrayFromImage(dilated_image)

# Convert the original image to an array for slicing
image_array = sitk.GetArrayFromImage(image)

# Prepare the plot
fig, axes = plt.subplots(1, 4, figsize=(20, 5))
plt.subplots_adjust(bottom=0.25)

# Set the background color to black
fig.patch.set_facecolor('black')
for ax in axes:
    ax.set_facecolor('black')

# Create a colormap for visualization
cmap = plt.get_cmap('tab20')

# Initialize slice index
slice_index = image_array.shape[0] // 2  # Start from the middle slice

def update_plot(slice_index):
    # Clear previous images from axes
    for ax in axes:
        ax.clear()  # Clear the axis

    # Original image (skip background)
    axes[0].imshow(np.where(image_array[slice_index] == 0, np.nan, image_array[slice_index]), cmap=cmap)
    axes[0].set_title(f'Original Image (Slice {slice_index})', color='white')
    axes[0].axis('off')

    # Eroded image (skip background)
    axes[1].imshow(np.where(eroded_image_array[slice_index] == 0, np.nan, eroded_image_array[slice_index]), cmap=cmap)
    axes[1].set_title(f'Eroded Image (Slice {slice_index})', color='white')
    axes[1].axis('off')

    # Filled image (skip background)
    axes[2].imshow(np.where(filled_image_array[slice_index] == 0, np.nan, filled_image_array[slice_index]), cmap=cmap)
    axes[2].set_title(f'Filled Image (Slice {slice_index})', color='white')
    axes[2].axis('off')

    # Dilated image (skip background)
    axes[3].imshow(np.where(dilated_image_array[slice_index] == 0, np.nan, dilated_image_array[slice_index]), cmap=cmap)
    axes[3].set_title(f'Dilated Image (Slice {slice_index})', color='white')
    axes[3].axis('off')

    plt.draw()  # Redraw the figure

# Create a slider for slice selection
ax_slider = plt.axes([0.2, 0.1, 0.6, 0.03], facecolor='lightgoldenrodyellow')
slice_slider = Slider(ax_slider, 'Slice', 0, image_array.shape[0] - 1, valinit=slice_index, valstep=1)

# Update the plot when the slider value changes
def update(val):
    update_plot(int(slice_slider.val))

slice_slider.on_changed(update)

# Initial plot
update_plot(slice_index)

# Show the plot in a pop-up window
plt.show()

In [6]:
# APARTE SLICES _ GEEN SLIDER
'''
# Perform binary erosion
# eroded_image = sitk.BinaryErode(image, kernelRadius=[0, 1, 0], foregroundValue=1)
# eroded_image_array = sitk.GetArrayFromImage(eroded_image)

erode_filter = sitk.BinaryErodeImageFilter() 
erode_filter.SetForegroundValue(1) 
erode_filter.SetKernelType(sitk.sitkBall) 
eroded_image = erode_filter.Execute(image)
eroded_image_array = sitk.GetArrayFromImage(eroded_image)

# Fill holes in the eroded image
# filled_image = sitk.BinaryFillhole(eroded_image)
fillhole_filter = sitk.BinaryFillholeImageFilter() 
fillhole_filter.SetFullyConnected(True) 
fillhole_filter.SetForegroundValue(1) 
filled_image = fillhole_filter.Execute(eroded_image) 
filled_image_array = sitk.GetArrayFromImage(filled_image)

# Perform binary dilation on the filled image
# dilated_image = sitk.BinaryDilate(filled_image, kernelRadius=[1, 1, 1])
dilate_filter = sitk.BinaryDilateImageFilter() 
dilate_filter.SetForegroundValue(1) 
dilate_filter.SetKernelType(sitk.sitkBall)
dilated_image = dilate_filter.Execute(filled_image) 
dilated_image_array = sitk.GetArrayFromImage(dilated_image)

# TRY TO USE CONNECTED COMPONENTS ON EACH INDIVUAL LABEL OF THE MASK AND AFTER COMBINE AGAIN


# Determine the slices to visualize (if the image is 3D)
if len(image_array.shape) == 3:
    slice_indices = [image_array.shape[0] // 2 - 10, 
                     image_array.shape[0] // 2,  # 2nd slice (middle)
                     image_array.shape[0] // 2 + 10]
    
    # Prepare the plots
    original_slices = [image_array[i, :, :] for i in slice_indices]
    eroded_slices = [eroded_image_array[i, :, :] for i in slice_indices]
    filled_slices = [filled_image_array[i, :, :] for i in slice_indices]
    dilated_slices = [dilated_image_array[i, :, :] for i in slice_indices]
else:
    original_slices = [image_array]
    eroded_slices = [eroded_image_array]
    filled_slices = [filled_image_array]
    dilated_slices = [dilated_image_array]

# Create a colormap for visualization
cmap = plt.colormaps.get_cmap('tab20')

# Set up the plot
fig, axes = plt.subplots(len(slice_indices), 4, figsize=(20, 15))

# Loop through the slices and plot
for i, idx in enumerate(slice_indices):
    # Original image (skip background)
    axes[i, 0].imshow(np.where(original_slices[i] == 0, np.nan, original_slices[i]), cmap=cmap)
    axes[i, 0].set_title(f'Original Image (Slice {idx})')
    axes[i, 0].axis('off')

    # Eroded image (skip background)
    axes[i, 1].imshow(np.where(eroded_slices[i] == 0, np.nan, eroded_slices[i]), cmap=cmap)
    axes[i, 1].set_title(f'Eroded Image (Slice {idx})')
    axes[i, 1].axis('off')

    # Filled image (skip background)
    axes[i, 2].imshow(np.where(filled_slices[i] == 0, np.nan, filled_slices[i]), cmap=cmap)
    axes[i, 2].set_title(f'Filled Image (Slice {idx})')
    axes[i, 2].axis('off')

    # Dilated image (skip background)
    axes[i, 3].imshow(np.where(dilated_slices[i] == 0, np.nan, dilated_slices[i]), cmap=cmap)
    axes[i, 3].set_title(f'Dilated Image (Slice {idx})')
    axes[i, 3].axis('off')

# Show the plot in a pop-up window
plt.tight_layout()
plt.show()
'''

"\n# Perform binary erosion\n# eroded_image = sitk.BinaryErode(image, kernelRadius=[0, 1, 0], foregroundValue=1)\n# eroded_image_array = sitk.GetArrayFromImage(eroded_image)\n\nerode_filter = sitk.BinaryErodeImageFilter() \nerode_filter.SetForegroundValue(1) \nerode_filter.SetKernelType(sitk.sitkBall) \neroded_image = erode_filter.Execute(image)\neroded_image_array = sitk.GetArrayFromImage(eroded_image)\n\n# Fill holes in the eroded image\n# filled_image = sitk.BinaryFillhole(eroded_image)\nfillhole_filter = sitk.BinaryFillholeImageFilter() \nfillhole_filter.SetFullyConnected(True) \nfillhole_filter.SetForegroundValue(1) \nfilled_image = fillhole_filter.Execute(eroded_image) \nfilled_image_array = sitk.GetArrayFromImage(filled_image)\n\n# Perform binary dilation on the filled image\n# dilated_image = sitk.BinaryDilate(filled_image, kernelRadius=[1, 1, 1])\ndilate_filter = sitk.BinaryDilateImageFilter() \ndilate_filter.SetForegroundValue(1) \ndilate_filter.SetKernelType(sitk.sitkBall)\n

### Getting the seperate label masks

In [7]:
labels = sitk.GetArrayFromImage(image).astype(int)
unique_labels = np.unique(labels)

In [8]:
unique_labels.shape

(6,)

In [9]:
'''
cmap = plt.colormaps.get_cmap('tab20')

# Get unique labels, excluding the background
unique_labels = np.unique(image_array)
unique_labels = unique_labels[unique_labels != 0]  # Ignore background because it's 0

# Select a specific slice index, for example, the middle slice along the z-axis
slice_index = image_array.shape[0] // 2

# Set up the subplots grid
num_labels = len(unique_labels)
fig, axes = plt.subplots(1, num_labels, figsize=(4 * num_labels, 4))
fig.suptitle("Individual Label Masks")

# Plot each label as a separate binary mask on the selected slice
for i, label in enumerate(unique_labels):
    label_mask = (image_array == label).astype(int)  # Create binary mask for each label
    ax = axes[i] if num_labels > 1 else axes  # Handle case with only one subplot
    ax.imshow(label_mask[slice_index, :, :], cmap="gray")  # Plot only the selected 2D slice
    ax.set_title(f"Label {label}")
    ax.axis("off")

plt.tight_layout()
plt.show()
'''

'\ncmap = plt.colormaps.get_cmap(\'tab20\')\n\n# Get unique labels, excluding the background\nunique_labels = np.unique(image_array)\nunique_labels = unique_labels[unique_labels != 0]  # Ignore background because it\'s 0\n\n# Select a specific slice index, for example, the middle slice along the z-axis\nslice_index = image_array.shape[0] // 2\n\n# Set up the subplots grid\nnum_labels = len(unique_labels)\nfig, axes = plt.subplots(1, num_labels, figsize=(4 * num_labels, 4))\nfig.suptitle("Individual Label Masks")\n\n# Plot each label as a separate binary mask on the selected slice\nfor i, label in enumerate(unique_labels):\n    label_mask = (image_array == label).astype(int)  # Create binary mask for each label\n    ax = axes[i] if num_labels > 1 else axes  # Handle case with only one subplot\n    ax.imshow(label_mask[slice_index, :, :], cmap="gray")  # Plot only the selected 2D slice\n    ax.set_title(f"Label {label}")\n    ax.axis("off")\n\nplt.tight_layout()\nplt.show()\n'

In [10]:
# WITH SLIDER

# Convert the SimpleITK image to a NumPy array
image_array = sitk.GetArrayFromImage(image)

# Get unique labels, excluding the background (assuming it's label 0)
unique_labels = np.unique(image_array)
unique_labels = unique_labels[unique_labels != 0]  # Ignore background if it's 0

# Set up the colormap
cmap = plt.get_cmap('tab20')

# Initialize the figure and axes
num_labels = len(unique_labels)
fig, axes = plt.subplots(1, num_labels, figsize=(4 * num_labels, 4))
fig.suptitle("Individual Label Masks")

# Set initial slice index
slice_index = image_array.shape[0] // 2

# Function to update the plot based on the selected slice index
def update_plot(slice_index):
    for ax in axes:
        ax.clear()  # Clear current axes
    # Plot each label as a separate binary mask on the selected slice
    for i, label in enumerate(unique_labels):
        label_mask = (image_array == label).astype(int)  # Create binary mask for each label
        ax = axes[i] if num_labels > 1 else axes  # Handle case with only one subplot
        ax.imshow(label_mask[slice_index, :, :], cmap="gray")  # Plot only the selected 2D slice
        ax.set_title(f"Label {label}")
        ax.axis("off")
    plt.draw()  # Redraw the figure

# Create the slider
ax_slider = plt.axes([0.1, 0.01, 0.8, 0.03])  # Position for the slider
slice_slider = Slider(ax_slider, 'Slice Index', 0, image_array.shape[0] - 1, valinit=slice_index, valstep=1)

# Connect the slider to the update function
slice_slider.on_changed(update_plot)

# Initial plot with the first slice
update_plot(slice_index)

# Show the plot
plt.show()

### Try only to do NComponents first

[Source code for Pymia PostProcessing](https://pymia.readthedocs.io/en/latest/_modules/pymia/filtering/postprocessing.html#BinaryThreshold): BinaryThreshold and LargestNConnectedComponents

In [51]:
# Assuming `image` is the original SimpleITK image
# Convert the SimpleITK image to a NumPy array for easier label extraction
image_array = sitk.GetArrayFromImage(image)
unique_labels = np.unique(image_array)
unique_labels = unique_labels[unique_labels != 0]  # Exclude the background (label 0)

# Initialize an empty image to accumulate processed results
processed_image = sitk.Image(image.GetSize(), sitk.sitkUInt8)
processed_image.CopyInformation(image)

# Initialize the connected components filter
largest_components_filter = pymia_postproc.LargestNConnectedComponents(number_of_components=1, consecutive_component_labels=True)

# Create a pipeline and add the filter to it
pipeline = pymia_fltr.FilterPipeline()
pipeline.add_filter(largest_components_filter)

# Process each label individually
for label in unique_labels:
    # Create a binary mask for the current label
    label_mask = sitk.BinaryThreshold(image, lowerThreshold=float(label), upperThreshold=float(label), insideValue=1, outsideValue=0)
    
    # Apply the pipeline to the binary mask
    processed_mask = pipeline.execute(label_mask)
    
    # Cast processed_mask to 8-bit and multiply by label
    processed_mask = sitk.Cast(sitk.Multiply(processed_mask, float(label)), sitk.sitkUInt8)
    
    # Add the processed mask to the final processed image
    processed_image = sitk.Add(processed_image, processed_mask)

# Convert processed image to NumPy array for plotting
processed_image_array = sitk.GetArrayFromImage(processed_image)
num_labels = len(unique_labels)
max_slice = processed_image_array.shape[0] - 1

# Set up the initial plot
fig, axes = plt.subplots(1, num_labels, figsize=(4 * num_labels, 4))
fig.suptitle("Processed Individual Label Masks")

# Initialize the initial slice index to the middle slice
initial_slice_index = max_slice // 2

# Display the initial slice for each label
def display_slice(slice_index):
    for i, label in enumerate(unique_labels):
        processed_label_mask = (processed_image_array == label).astype(int)
        ax = axes[i] if num_labels > 1 else axes
        ax.clear()
        ax.imshow(processed_label_mask[slice_index, :, :], cmap="gray")
        ax.set_title(f"Processed Label {label}")
        ax.axis("off")

    fig.canvas.draw_idle()

display_slice(initial_slice_index)

# Add a slider below the plots
ax_slider = plt.axes([0.2, 0.05, 0.6, 0.03], facecolor='lightgoldenrodyellow')
slice_slider = Slider(ax_slider, 'Slice', 0, max_slice, valinit=initial_slice_index, valstep=1)

# Update the displayed slice when the slider is changed
def update(val):
    slice_index = int(slice_slider.val)
    display_slice(slice_index)

slice_slider.on_changed(update)

plt.show()

In [52]:
# Assuming processed_image_array and unique_labels are already defined
num_labels = len(unique_labels)
max_slice = processed_image_array.shape[0] - 1

# Create a color map for the labels using 'tab20'
cmap = plt.get_cmap('tab20')  # Use 'tab20' for distinct colors

# Set up the initial plot
fig, ax = plt.subplots(figsize=(8, 8))
fig.suptitle("Processed Label Masks")

# Initialize the initial slice index to the middle slice
initial_slice_index = max_slice // 2

# Function to display the whole mask with different colors
def display_slice(slice_index):
    rgb_image = np.zeros((processed_image_array.shape[1], processed_image_array.shape[2], 3), dtype=np.uint8)
    for i, label in enumerate(unique_labels):
        # Create a mask for the current label
        mask = (processed_image_array[slice_index, :, :] == label)
        # Convert color tuple to a NumPy array and assign color only to the pixels where the mask is True
        color = np.array(cmap(i)[:3]) * 255  # Convert to array and scale to [0, 255]
        rgb_image[mask] = color.astype(np.uint8)  # Assign color to masked pixels
    ax.clear()
    ax.imshow(rgb_image)
    ax.set_title(f"Processed Labels at Slice {slice_index}")
    ax.axis("off")
    fig.canvas.draw_idle()

display_slice(initial_slice_index)

# Add a slider below the plots
ax_slider = plt.axes([0.2, 0.05, 0.6, 0.03], facecolor='lightgoldenrodyellow')
slice_slider = Slider(ax_slider, 'Slice', 0, max_slice, valinit=initial_slice_index, valstep=1)

# Update the displayed slice when the slider is changed
def update(val):
    slice_index = int(slice_slider.val)
    display_slice(slice_index)

slice_slider.on_changed(update)

plt.show()

### N_comp + pipeline

In [53]:
# Initialize an empty image to accumulate processed results
processed_image_ = sitk.Image(image.GetSize(), sitk.sitkUInt8)
processed_image_.CopyInformation(image)

# initialize parameters
largest_components_filter = pymia_postproc.LargestNConnectedComponents(number_of_components=1, consecutive_component_labels=False)
print(largest_components_filter)

# make pipeline of postproc filters
pipeline = pymia_fltr.FilterPipeline()
pipeline.add_filter(largest_components_filter)

# apply pipeline
processed_image_ = pipeline.execute(image)
processed_image_array_ = sitk.GetArrayFromImage(processed_image_)


# Define the image processing pipeline
erode_filter = sitk.BinaryErodeImageFilter() 
erode_filter.SetForegroundValue(1) 
erode_filter.SetKernelType(sitk.sitkBall) 
eroded_image_ = erode_filter.Execute(processed_image)
eroded_image_array_ = sitk.GetArrayFromImage(eroded_image_)

fillhole_filter = sitk.BinaryFillholeImageFilter() 
fillhole_filter.SetFullyConnected(True) 
fillhole_filter.SetForegroundValue(1) 
filled_image_ = fillhole_filter.Execute(eroded_image_) 
filled_image_array_ = sitk.GetArrayFromImage(filled_image_)

dilate_filter = sitk.BinaryDilateImageFilter() 
dilate_filter.SetForegroundValue(1) 
dilate_filter.SetKernelType(sitk.sitkBall)
dilated_image_ = dilate_filter.Execute(filled_image_) 
combined_pipeline = sitk.GetArrayFromImage(dilated_image_)

LargestNConnectedComponents:
 number_of_components:         1
 consecutive_component_labels: False



# Compare with GT

In [54]:
def compute_metrics(predicted_mask, gt_mask):
    # Flatten the masks
    predicted_mask_flat = predicted_mask.flatten()
    gt_mask_flat = gt_mask.flatten()

    # Calculate Intersection and Union
    intersection = np.sum(predicted_mask_flat * gt_mask_flat)
    union = np.sum(predicted_mask_flat) + np.sum(gt_mask_flat) - intersection

    # Calculate metrics
    iou = intersection / union if union != 0 else 0
    dice = (2 * intersection) / (np.sum(predicted_mask_flat) + np.sum(gt_mask_flat)) if (np.sum(predicted_mask_flat) + np.sum(gt_mask_flat)) != 0 else 0
    accuracy = np.sum(predicted_mask_flat == gt_mask_flat) / gt_mask_flat.size
    true_positive = np.sum((predicted_mask_flat == 1) & (gt_mask_flat == 1))
    false_positive = np.sum((predicted_mask_flat == 1) & (gt_mask_flat == 0))
    false_negative = np.sum((predicted_mask_flat == 0) & (gt_mask_flat == 1))
    precision = true_positive / (true_positive + false_positive) if (true_positive + false_positive) != 0 else 0
    recall = true_positive / (true_positive + false_negative) if (true_positive + false_negative) != 0 else 0

    return iou, dice, accuracy, precision, recall

## Original

In [55]:
# Assuming GT_image_array and image_array are already defined

# Create binary masks from the ground truth and original image
gt_mask = (GT_array > 0).astype(np.uint8)  # Ground truth mask
original_mask = (image_array > 0).astype(np.uint8)  # Original image mask

# Compute overall metrics between the original image and ground truth
iou_original_gt, dice_original_gt, accuracy_original_gt, precision_original_gt, recall_original_gt = compute_metrics(original_mask, gt_mask)

# Print overall metrics for original image vs. ground truth
print("Overall Metrics (Original Image vs. Ground Truth):")
print(f"IoU: {iou_original_gt:.4f}, Dice: {dice_original_gt:.4f}, Accuracy: {accuracy_original_gt:.4f}, Precision: {precision_original_gt:.4f}, Recall: {recall_original_gt:.4f}")


Overall Metrics (Original Image vs. Ground Truth):
IoU: 0.6349, Dice: 0.7767, Accuracy: 0.9334, Precision: 0.6377, Recall: 0.9929


## Pipeline

In [56]:
# Get binary masks from images (assuming foreground value is 1)
eroded_mask = (eroded_image_array > 0).astype(np.uint8)
filled_mask = (filled_image_array > 0).astype(np.uint8)
dilated_mask = (dilated_image_array > 0).astype(np.uint8)
gt_mask = (GT_array > 0).astype(np.uint8)
original_mask = (image_array > 0).astype(np.uint8)  # Original image as binary mask

# Compute metrics for each processing step
iou_eroded, dice_eroded, accuracy_eroded, precision_eroded, recall_eroded = compute_metrics(eroded_mask, gt_mask)
iou_filled, dice_filled, accuracy_filled, precision_filled, recall_filled = compute_metrics(filled_mask, gt_mask)
iou_dilated, dice_dilated, accuracy_dilated, precision_dilated, recall_dilated = compute_metrics(dilated_mask, gt_mask)

# Print metrics
print("\nEroded Mask Metrics:")
print(f"IoU: {iou_eroded:.4f}, Dice: {dice_eroded:.4f}, Accuracy: {accuracy_eroded:.4f}, Precision: {precision_eroded:.4f}, Recall: {recall_eroded:.4f}")

print("\nFilled Mask Metrics:")
print(f"IoU: {iou_filled:.4f}, Dice: {dice_filled:.4f}, Accuracy: {accuracy_filled:.4f}, Precision: {precision_filled:.4f}, Recall: {recall_filled:.4f}")

print("\nDilated Mask Metrics:")
print(f"IoU: {iou_dilated:.4f}, Dice: {dice_dilated:.4f}, Accuracy: {accuracy_dilated:.4f}, Precision: {precision_dilated:.4f}, Recall: {recall_dilated:.4f}")


Eroded Mask Metrics:
IoU: 0.6024, Dice: 0.7519, Accuracy: 0.9351, Precision: 0.6782, Recall: 0.8435

Filled Mask Metrics:
IoU: 0.6024, Dice: 0.7518, Accuracy: 0.9351, Precision: 0.6782, Recall: 0.8435

Dilated Mask Metrics:
IoU: 0.6366, Dice: 0.7780, Accuracy: 0.9359, Precision: 0.6524, Recall: 0.9635


## N_components

In [57]:
# Assuming processed_image_array, GT_image_array, and image_array are already defined

# Create binary masks from the processed image, ground truth, and original image
processed_mask = (processed_image_array > 0).astype(np.uint8)  # Combine all labels in the processed image
gt_mask = (GT_array > 0).astype(np.uint8)  # Ground truth mask
original_mask = (image_array > 0).astype(np.uint8)  # Original image mask

# Compute overall metrics against the ground truth
iou_gt, dice_gt, accuracy_gt, precision_gt, recall_gt = compute_metrics(processed_mask, gt_mask)

# Compute overall metrics against the original image
iou_original, dice_original, accuracy_original, precision_original, recall_original = compute_metrics(processed_mask, original_mask)

# Print overall metrics for ground truth
print("vs. Ground Truth:")
print(f"IoU: {iou_gt:.4f}, Dice: {dice_gt:.4f}, Accuracy: {accuracy_gt:.4f}, Precision: {precision_gt:.4f}, Recall: {recall_gt:.4f}")

# Print overall metrics for original image
print("\nvs. Original Image:")
print(f"IoU: {iou_original:.4f}, Dice: {dice_original:.4f}, Accuracy: {accuracy_original:.4f}, Precision: {precision_original:.4f}, Recall: {recall_original:.4f}")

vs. Ground Truth:
IoU: 0.6380, Dice: 0.7790, Accuracy: 0.9344, Precision: 0.6418, Recall: 0.9910

vs. Original Image:
IoU: 0.9918, Dice: 0.9959, Accuracy: 0.9985, Precision: 1.0000, Recall: 0.9918


For N=2 (worse):

IoU: 0.6378, Dice: 0.7788, Accuracy: 0.9344, Precision: 0.6414, Recall: 0.9911

## N_comp + pipeline

In [58]:
# Get binary masks from images (assuming foreground value is 1)
processed_mask_ = (processed_image_array_ > 0).astype(np.uint8)  # Combine all labels in the processed image

eroded_mask_ = (eroded_image_array_ > 0).astype(np.uint8)
filled_mask_ = (filled_image_array_ > 0).astype(np.uint8)
dilated_mask_ = (combined_pipeline > 0).astype(np.uint8)
gt_mask = (GT_array > 0).astype(np.uint8)
original_mask = (image_array > 0).astype(np.uint8)  # Original image as binary mask

# Compute metrics for each processing step
iou_proc, dice_proc, accuracy_proc, precision_proc, recall_proc = compute_metrics(processed_mask_, gt_mask)
iou_eroded, dice_eroded, accuracy_eroded, precision_eroded, recall_eroded = compute_metrics(eroded_mask_, gt_mask)
iou_filled, dice_filled, accuracy_filled, precision_filled, recall_filled = compute_metrics(filled_mask_, gt_mask)
iou_dilated, dice_dilated, accuracy_dilated, precision_dilated, recall_dilated = compute_metrics(dilated_mask_, gt_mask)

# Print metrics
print("\nProcessed Mask Metrics:")
print(f"IoU: {iou_proc:.4f}, Dice: {dice_proc:.4f}, Accuracy: {accuracy_proc:.4f}, Precision: {precision_proc:.4f}, Recall: {recall_proc:.4f}")

print("\nEroded Mask Metrics:")
print(f"IoU: {iou_eroded:.4f}, Dice: {dice_eroded:.4f}, Accuracy: {accuracy_eroded:.4f}, Precision: {precision_eroded:.4f}, Recall: {recall_eroded:.4f}")

print("\nFilled Mask Metrics:")
print(f"IoU: {iou_filled:.4f}, Dice: {dice_filled:.4f}, Accuracy: {accuracy_filled:.4f}, Precision: {precision_filled:.4f}, Recall: {recall_filled:.4f}")

print("\nDilated Mask Metrics:")
print(f"IoU: {iou_dilated:.4f}, Dice: {dice_dilated:.4f}, Accuracy: {accuracy_dilated:.4f}, Precision: {precision_dilated:.4f}, Recall: {recall_dilated:.4f}")


Processed Mask Metrics:
IoU: 0.6352, Dice: 0.7769, Accuracy: 0.9335, Precision: 0.6381, Recall: 0.9929

Eroded Mask Metrics:
IoU: 0.6039, Dice: 0.7530, Accuracy: 0.9355, Precision: 0.6804, Recall: 0.8430

Filled Mask Metrics:
IoU: 0.6039, Dice: 0.7530, Accuracy: 0.9355, Precision: 0.6804, Recall: 0.8430

Dilated Mask Metrics:
IoU: 0.6382, Dice: 0.7791, Accuracy: 0.9363, Precision: 0.6542, Recall: 0.9631
